sel/src/Sel/Hashing/Password.hs

```diff
@@ -0,0 +1,206 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- |
+--
+-- Module: Sel.Hashing.Password
+-- Description: Password hashing with the Argon2id algorithm
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Portability: GHC only
+module Sel.Hashing.Password
+  ( -- * Introduction
+    -- $introduction
+    PasswordHash
+
+    -- ** Password Hashing and Verifying
+  , hashPassword
+  , verifyPassword
+  , hashPasswordWithParams
+
+    -- ** Conversion to textual formats
+  , passwordHashToByteString
+  , passwordHashToText
+
+    -- * Salt
+  , Salt (..)
+  , genSalt
+
+    -- * Argon2 Parameters
+  , Argon2Params (..)
+  , defaultArgon2Params
+  )
+where
+
+import Control.Monad (void)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Internal as ByteString
+import qualified Data.ByteString.Unsafe as ByteString
+import Data.Text (Text)
+import Data.Text.Display
+import qualified Data.Text.Encoding as Text
+import qualified Data.Text.Foreign as Text
+import qualified Data.Text.Lazy.Builder as Builder
+import Foreign hiding (void)
+import Foreign.C
+import GHC.Generics
+import GHC.IO.Handle.Text (memcpy)
+import LibSodium.Bindings.PasswordHashing
+import LibSodium.Bindings.Random
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+-- $introduction
+--
+-- This API provides functions for password hashing, backed by the Argon2id algorithm.
+--
+-- If you need to deviate from the defaults enforced by this module,
+-- please use the underlying bindings at "LibSodium.Bindings.PasswordHashing".
+
+-- |
+--
+-- @since 0.0.1.0
+newtype PasswordHash = PasswordHash (ForeignPtr CChar)
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    , Show
+      -- ^ @since 0.0.1.0
+    )
+
+-- | @since 0.0.1.0
+instance Display PasswordHash where
+  displayBuilder = Builder.fromText . passwordHashToText
+
+-- | Hash the password with the Argon2id algorithm and
+-- a set of pre-defined parameters
+--
+-- @since 0.0.1.0
+hashPassword :: Text -> IO PasswordHash
```

Comment:
I'd rather accept `ByteString` here, with possibly a `Text` helper. Your data encoding is not our concern.